docs: add 'Migrate from Gen1' page and severity classification (#6275)
adobe/spectrum-web-components: 2nd-gen/packages/swc/.storybook/resources/migrate-from-gen1.mdx

import { Meta } from '@storybook/addon-docs/blocks';

<Meta title="Migrate from Gen1" />

# Migrate from Gen1

This page walks an application team through moving from SWC Gen1 (`@spectrum-web-components/*`) to SWC Gen2 (`@adobe/spectrum-wc`). You do not have to migrate everything at once — Gen1 and Gen2 are designed to coexist in the same app so you can move one component (or one screen) at a time.

For a high-level comparison of what changed, see [Gen1 vs Gen2](/docs/learn-about-swc-gen1-vs-gen2--docs).

## TL;DR

1. Install `@adobe/spectrum-wc` alongside `@spectrum-web-components/*`.

> 2. Switch Gen1 to the `spectrum-two` theme so it matches Gen2 visually.

Reviewer: Hmm, this one is a little tricky... maybe:

```suggestion
2. Switch Gen1 to the `spectrum-two` theme so it better aligns with Gen2 visually.
```

Author: I rewrote this 3 times and none were as good as your suggestion 😅

3. Add the Gen2 stylesheet and `swc-theme` classes to your application root.
4. Load Adobe Clean Spectrum VF (the Spectrum 2 font family).

> 5. Replace `<sp-*>` with `<swc-*>` one component at a time, following each component's **Migration guide**.

Reviewer 2: Should we link to the docs and call out that some components are still in progress? For example, Illustrated Message currently omits the button group while that dependency is being built out. Worth flagging so teams can check component readiness before committing to a migration.

Author: Yes! I now reference the "Upcoming features"

> 6. Swap `--mod-*` overrides for `--swc-*` overrides as you go.

Reviewer: We don't want to imply this is _required_ especially as these are not 1:1 in intent or availability.

```suggestion
6. Remove deprecated `--mod-*` overrides and apply `--swc-*` overrides if necessary.
```

7. Once no `<sp-*>` elements remain, remove the Gen1 package, the `spectrum-two` theme imports, and the `<sp-theme>` element.

## Gen1 and Gen2 can coexist

Three properties make a gradual migration possible:

- **Different package names** — `@spectrum-web-components` and `@adobe/spectrum-wc` install side by side without conflicting.
- **Different tag prefixes** — `<sp-badge>` and `<swc-badge>` are independent custom elements. You can have both in the same tree.

> - **Shared visual language** — Gen1 includes a `spectrum-two` theme that serves as a stepping stone to the Spectrum 2 design system Gen2 implements natively. Switch Gen1 to that theme so the two generations look cohesive on the page.

Reviewer: ```suggestion
- **Shared visual language** — Gen1 includes a `spectrum-two` theme that serves as a stepping stone to the Spectrum 2 design system Gen2 implements natively. Switch Gen1 to that theme so the two generations look cohesive on the page.
- **Separate token namespace** — Design token custom properties use the `swc` namespace to prevent conflicts with Gen1 visuals.
```

Reviewer 3: ```suggestion
- **Shared visual language** — Gen1's`spectrum-two` theme serves as a bridge between Spectrum/Express and Spectrum 2, which Gen2 uses out of the box. Using this theme on Gen1 components creates visual consistency between both generations.
```

## Step 1 — Install Gen2

```bash
yarn add @adobe/spectrum-wc
```

Gen2 ships as a single package; you import individual components from subpaths. You can keep your existing `@spectrum-web-components/*` packages.

## Step 2 — Align Gen1 visually with Spectrum 2

> Switch your existing `<sp-theme>` to the `spectrum-two` system so any remaining Gen1 components match Gen2 visually during the migration:

Reviewer: ```suggestion
Switch your existing `<sp-theme>` to the `spectrum-two` system so any remaining Gen1 components align with Gen2 visually during the migration:
```

```ts
import '@spectrum-web-components/theme/sp-theme.js';
import '@spectrum-web-components/theme/src/spectrum-two/themes.js';
```

```html
<sp-theme system="spectrum-two" color="light" scale="medium">
  <!-- existing sp-* components -->
</sp-theme>
```

> Gen1 components inside this theme adopt Spectrum 2 tokens and look the same as `<swc-*>` neighbors.

Reviewer: ```suggestion
Gen1 components inside this theme adopt Spectrum 2 tokens and look cohesive alongside `<swc-*>` neighbors.
```

> Gen1 components inside this theme adopt Spectrum 2 tokens and look the same as `<swc-*>` neighbors.

Reviewer 3: ```suggestion
Gen1 components inside this theme adopt Spectrum 2 tokens and use the same colours, border radii, and icons as their `<swc-*>` neighbors.
```

## Step 4 — Load Adobe Clean Spectrum VF

Gen2 uses Adobe Clean Spectrum VF, the Spectrum 2 font family. It looks visually identical to Adobe Clean (the Gen1 font) but ships updated metrics for better vertical alignment inside components.

How you load it depends on your environment — for Adobe Fonts (Typekit), CDN, or self-hosted setups, see the [Fonts guide](/docs/guides-customization-fonts--docs).

## Step 5 — Replace components one at a time

For each migrated Gen2 component:

1. Open the component's **Migration guide** in this site (e.g. [Badge](/docs/components-badge-migration-guide--docs)). It enumerates the breaking attribute, slot, event, and default-value changes.
2. Update the import to the per-element side-effect path:

   ```ts
   // Before
   import '@spectrum-web-components/badge/sp-badge.js';

   // After
   import '@adobe/spectrum-wc/components/badge/swc-badge.js';
   ```

3. Rename the tag (`<sp-badge>` → `<swc-badge>`) and apply any attribute renames called out in the migration guide.

> 4. Move CSS overrides from `--mod-*` to the equivalent `--swc-*` properties (covered below).

Reviewer: Again, I don't want to imply this is required. Hopefully S2 tokens reduce this need almost entirely, outside of maybe product specific colors!

```suggestion
4. Remove deprecated `--mod-*` overrides and apply `--swc-*` overrides if necessary (covered below).
```

The components index on this site lists every Gen2 component that has shipped, with its Migration guide and full API reference.

## Step 6 — Migrate styling

Two things change for application CSS.

### `--mod-*` → `--swc-*`

> Gen2 collapses the `--mod-*` custom-property API into a smaller, namespaced `--swc-*` API. Common rewrites:

Reviewer: ```suggestion
Gen2 deprecates the `--mod-*` custom property API into a smaller, namespaced `--swc-*` API that is not guaranteed to be a 1:1 replacement due to Spectrum 2 aligned style updates. We suggest holding off on rewriting previous `--mod-*` overrides until you examine the component's appearance in your application.

Example rewrites:
```

```css
/* Before */
sp-badge {
  --mod-badge-height: 24px;
  --mod-badge-corner-radius: 4px;
  --mod-badge-label-icon-color: var(--spectrum-gray-900);
}

/* After */
swc-badge {
  --swc-badge-height: 24px;
  --swc-badge-corner-radius: 4px;
  --swc-badge-label-icon-color: var(--swc-gray-900);
}
```

> Each per-component Migration guide includes a mapping table for that component's properties. The general namespacing rule: `--mod-` becomes `--swc-`, and `--spectrum-` (token references) also become `--swc-` because Gen2 republishes Spectrum 2 tokens under its own namespace.

Reviewer 3: nit
```suggestion
Each per-component migration guide includes a mapping table for that component's properties. The general namespacing rule: `--mod-` becomes `--swc-`, and `--spectrum-` (token references) also become `--swc-` because Gen2 republishes Spectrum 2 tokens under its own namespace.
```

### `<sp-theme>` → CSS classes

Anything you were doing with `<sp-theme>` (theme, scale, language, direction) now lives on a root element via classes and standard HTML attributes:

| Gen1 (`<sp-theme>`)        | Gen2 (root element)                                       |
| -------------------------- | --------------------------------------------------------- |
| `color="light"` / `"dark"` | `class="swc-theme swc-theme--light"` / `swc-theme--dark`  |
| `scale="medium"` / `large` | `class="swc-theme swc-theme--sizeM"` / `swc-theme--sizeL` |
| `lang="ja"` (on element)   | `lang="ja"` on `<html>` or the scoping element            |
| `dir="rtl"`                | `dir="rtl"` on `<html>` or the scoping element            |

You can keep the Gen1 `<sp-theme>` wrapping any remaining `<sp-*>` components for as long as you need; the two systems do not conflict.

## Step 7 — Decommission Gen1

Once no `<sp-*>` element remains in your app:

1. Remove `@spectrum-web-components/*` packages from your `package.json`.
2. Remove the `<sp-theme>` element and its imports (`@spectrum-web-components/theme/*`).
3. Remove any remaining `--mod-*` overrides in your CSS.

## What's not migrated yet

> Gen2 is shipping components incrementally. The components index in this site lists every component that has a Gen2 implementation; anything not listed is still Gen1 only. Plan to keep the Gen1 package installed alongside Gen2 until your components of interest are migrated.

Reviewer 3: ```suggestion
Gen2 is shipping components incrementally. The index of components on our docs site lists all components with a Gen2 implementation; anything not listed is still Gen1 only. Plan to keep the Gen1 package installed alongside Gen2 until your components of interest are migrated.
```

> Gen2 is shipping components incrementally. The components index in this site lists every component that has a Gen2 implementation; anything not listed is still Gen1 only. Plan to keep the Gen1 package installed alongside Gen2 until your components of interest are migrated.

Reviewer 2: Should we add a note to the migration guide about bundle size implications during the dual-import phase Could be good to mention, along with some potential strategies, like you suggested in the office hours.

Author: Great call! I added a section with this
